fvm/transactionInvocator.go

```diff
@@ -0,0 +1,332 @@
+	"github.com/onflow/flow-go/fvm/programs"
+	"github.com/onflow/flow-go/fvm/state"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/trace"
+)
+
+type TransactionInvocator struct {
+	logger zerolog.Logger
+}
+
+func NewTransactionInvocator(logger zerolog.Logger) *TransactionInvocator {
+	return &TransactionInvocator{
+		logger: logger,
+	}
+}
+
+func (i *TransactionInvocator) Process(
+	vm *VirtualMachine,
+	ctx *Context,
+	proc *TransactionProcedure,
+	sth *state.StateHolder,
+	programs *programs.Programs,
+) (txError errors.TransactionError, vmError errors.VMError) {
+
+	var span opentracing.Span
+	if ctx.Tracer != nil && proc.TraceSpan != nil {
+		span = ctx.Tracer.StartSpanFromParent(proc.TraceSpan, trace.FVMExecuteTransaction)
+		span.LogFields(
+			traceLog.String("transaction.ID", proc.ID.String()),
+		)
+		defer span.Finish()
+	}
+
+	var blockHeight uint64
+	if ctx.BlockHeader != nil {
+		blockHeight = ctx.BlockHeader.Height
+	}
+
+	var env *hostEnv
+	env = newEnvironment(*ctx, vm, sth, programs)
+	predeclaredValues := i.valueDeclarations(ctx, env)
+
+	retry := false
+	numberOfRetries := 0
+
+	parentState := sth.State()
+	childState := sth.NewChild()
+	defer func() {
+		if mergeError := parentState.MergeState(childState); mergeError != nil {
+			vmError = mergeError.(errors.VMError)
+			return
+		}
+		sth.SetActiveState(parentState)
+	}()
+
+	for numberOfRetries = 0; numberOfRetries < int(ctx.MaxNumOfTxRetries); numberOfRetries++ {
+		if retry {
+			// rest state
+			sth.SetActiveState(parentState)
+			childState = sth.NewChild()
+			// force cleanup if retries
+			programs.ForceCleanup()
+
+			i.logger.Warn().
+				Str("txHash", proc.ID.String()).
+				Uint64("blockHeight", blockHeight).
+				Int("retries_count", numberOfRetries).
+				Uint64("ledger_interaction_used", sth.State().InteractionUsed()).
+				Msg("retrying transaction execution")
+
+			// reset error part of proc
+			// Warning right now the tx requires retry logic doesn't change
+			// anything on state but we might want to revert the state changes (or not commiting)
+			// if we decided to expand it furthur.
+			proc.Err = nil
+			proc.Logs = make([]string, 0)
+			proc.Events = make([]flow.Event, 0)
+			proc.ServiceEvents = make([]flow.Event, 0)
+
+			// reset env
+			env = newEnvironment(*ctx, vm, sth, programs)
+		}
+
+		env.setTransaction(proc.Transaction, proc.TxIndex)
+		env.setTraceSpan(span)
+
+		location := common.TransactionLocation(proc.ID[:])
+
+		err := vm.Runtime.ExecuteTransaction(
+			runtime.Script{
+				Source:    proc.Transaction.Script,
+				Arguments: proc.Transaction.Arguments,
+			},
+			runtime.Context{
+				Interface:         env,
+				Location:          location,
+				PredeclaredValues: predeclaredValues,
+			},
+		)
+		if err != nil {
+			// TODO ExecuteTransaction returns both txErr and vmErr
+			txError, vmError = errors.HandleRuntimeError(err)
+			if vmError != nil {
+				return nil, vmError
+			}
```

Comment:
This is the only part that has been updated, the rest of the file are mostly the same as before just moved into txInvocator file